homeassistant/components/modbus/sensor.py

```diff
@@ -99,63 +112,89 @@ def number(value: Any) -> int | float:
+            structure = str(entry[CONF_COUNT] * 2) + "s"
+        elif entry[CONF_DATA_TYPE] != DATA_TYPE_CUSTOM:
             try:
-                structure = f">{DEFAULT_STRUCT_FORMAT[register[CONF_DATA_TYPE]][register[CONF_COUNT]]}"
+                structure = f">{DEFAULT_STRUCT_FORMAT[entry[CONF_DATA_TYPE]][entry[CONF_COUNT]]}"
             except KeyError:
                 _LOGGER.error(
                     "Unable to detect data type for %s sensor, try a custom type",
-                    register[CONF_NAME],
+                    entry[CONF_NAME],
                 )
                 continue
         else:
-            structure = register.get(CONF_STRUCTURE)
+            structure = entry.get(CONF_STRUCTURE)
 
         try:
             size = struct.calcsize(structure)
         except struct.error as err:
-            _LOGGER.error("Error in sensor %s structure: %s", register[CONF_NAME], err)
+            _LOGGER.error("Error in sensor %s structure: %s", entry[CONF_NAME], err)
             continue
 
-        if register[CONF_COUNT] * 2 != size:
+        if entry[CONF_COUNT] * 2 != size:
             _LOGGER.error(
                 "Structure size (%d bytes) mismatch registers count (%d words)",
                 size,
-                register[CONF_COUNT],
+                entry[CONF_COUNT],
             )
             continue
 
-        hub_name = register[CONF_HUB]
-        hub = hass.data[MODBUS_DOMAIN][hub_name]
+        if CONF_HUB in entry:
+            # from old config!
+            discovery_info[CONF_NAME] = entry[CONF_HUB]
+        if CONF_SCAN_INTERVAL not in entry:
+            entry[CONF_SCAN_INTERVAL] = DEFAULT_SCAN_INTERVAL
+        hub: ModbusHub = hass.data[MODBUS_DOMAIN][discovery_info[CONF_NAME]]
         sensors.append(
             ModbusRegisterSensor(
                 hub,
-                register[CONF_NAME],
-                register.get(CONF_SLAVE),
-                register[CONF_REGISTER],
-                register[CONF_REGISTER_TYPE],
-                register.get(CONF_UNIT_OF_MEASUREMENT),
-                register[CONF_COUNT],
-                register[CONF_REVERSE_ORDER],
-                register[CONF_SCALE],
-                register[CONF_OFFSET],
+                entry[CONF_NAME],
+                entry.get(CONF_SLAVE),
+                entry[CONF_ADDRESS],
+                entry[CONF_INPUT_TYPE],
+                entry.get(CONF_UNIT_OF_MEASUREMENT),
+                entry[CONF_COUNT],
+                entry[CONF_REVERSE_ORDER],
+                entry[CONF_SCALE],
+                entry[CONF_OFFSET],
                 structure,
-                register[CONF_PRECISION],
-                register[CONF_DATA_TYPE],
-                register.get(CONF_DEVICE_CLASS),
+                entry[CONF_PRECISION],
+                entry[CONF_DATA_TYPE],
+                entry.get(CONF_DEVICE_CLASS),
+                entry[CONF_SCAN_INTERVAL],
             )
         )
 
     if not sensors:
         return False
```

Comment:
Nothing is checking this return value.
```suggestion
        return
```